src/reducers/breakpoints.js

```diff
@@ -176,55 +155,101 @@ function addBreakpoint(state, action) {
 }
 
 function removeBreakpoint(state, action) {
-  if (action.status === "done") {
-    const id = makeLocationId(action.breakpoint.location);
+  if (action.status != "done") {
+    return state;
+  }
 
-    if (action.disabled) {
-      const bp = state.breakpoints.get(id);
-      const updatedState = state.setIn(
-        ["breakpoints", id],
-        updateObj(bp, {
-          loading: false,
-          disabled: true
-        })
-      );
+  const id = makeLocationId(action.breakpoint.location);
+  const pendingId = makePendingLocationId(action.breakpoint.location);
+  let updatedState = undefined;
+
+  if (action.disabled) {
+    const bp = state.breakpoints.get(id);
+    const breakpoint = updateObj(bp, {
+      loading: false,
+      disabled: true
+    });
+    updatedState = state.setIn(["breakpoints", id], breakpoint);
+    updatedState = updatePendingBreakpoint(updatedState, breakpoint);
+  } else {
+    updatedState = state
+      .deleteIn(["breakpoints", id])
+      .deleteIn(["pendingBreakpoints", pendingId]);
+  }
 
-      return updatedState.set(
-        "breakpointsDisabled",
-        allBreakpointsDisabled(updatedState)
-      );
-    }
+  return updatedState.set(
+    "breakpointsDisabled",
+    allBreakpointsDisabled(updatedState)
+  );
+}
 
-    const updatedState = state.deleteIn(["breakpoints", id]);
+function setCondition(state, action) {
+  const id = makeLocationId(action.breakpoint.location);
 
-    return updatedState.set(
-      "breakpointsDisabled",
-      allBreakpointsDisabled(updatedState)
+  if (action.status === "start") {
+    const bp = state.breakpoints.get(id);
+    return state.setIn(
+      ["breakpoints", id],
+      updateObj(bp, {
+        loading: true,
+        condition: action.condition
+      })
     );
   }
 
-  return state;
+  if (action.status === "done") {
+    const bp = state.breakpoints.get(id);
+    const updatedBreakpoint = updateObj(bp, {
+      id: action.value.id,
+      loading: false
+    });
+
+    let updatedState = state.setIn(["breakpoints", id], updatedBreakpoint);
+
+    return updatePendingBreakpoint(updatedState, updatedBreakpoint);
+  }
+
+  if (action.status === "error") {
+    return state.deleteIn(["breakpoints", id]);
+  }
+}
+
+function slideBreakpoint(state, action) {
+  const { actualLocation, id } = action.value;
+  const { breakpoint } = action;
+  const currentBp = state.breakpoints.get(id) || fromJS(breakpoint);
+
+  const locationId = makeLocationId(breakpoint.location);
+  const movedLocationId = makeLocationId(actualLocation);
+  state = state.deleteIn(["breakpoints", locationId]);
```

Comment:
Personal comment: Its not wrong, I just cringe if I see state being reassigned :P